File: cpp/150_evaluate_reverse_polish_notation.cpp
#include <stack>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 private:
  long Compute(long val1, long val2, const string &op) {
    switch (op[0]) {
      case '+':
        return val1 + val2;
      case '-':
        return val1 - val2;
      case '*':
        return val1 * val2;
      case '/':
        return val1 / val2;
    }
    return 0;
  }

 public:
  int evalRPN(vector<string> &tokens) {
    stack<long> stack;
    unordered_set<string> operators = {"+", "-", "*", "/"};
    for (const auto &token : tokens) {
      if (operators.count(token)) {
        long val2 = stack.top();
        stack.pop();
        long val1 = stack.top();
        stack.pop();
        long result = Compute(val1, val2, token);
        stack.push(result);
      } else {
        stack.push(stoi(token));
      }
    }
    return stack.top();
  }
};

File: rust/150_evaluate_reverse_polish_notation.rs
impl Solution {
    pub fn eval_rpn(tokens: Vec<String>) -> i32 {
        let mut stack = vec![];
        for token in &tokens {
            if let Ok(val) = token.parse() {
                stack.push(val);
                continue;
            }
            let rhs = stack.pop().unwrap();
            let lhs = stack.pop().unwrap();
            let result = match token.as_str() {
                "+" => lhs + rhs,
                "-" => lhs - rhs,
                "*" => lhs * rhs,
                "/" => lhs / rhs,
                _ => unreachable!(),
            };
            stack.push(result);
        }

        *stack.last().unwrap()
    }
}
